lib/Zonemaster/Backend/DB.pm

```diff
@@ -452,24 +583,110 @@ sub batch_exists_in_db {
     return $id;
 }
 
+=head2 get_test_request( $queue_label )
+
+Find a waiting test and claim it for processing.
+
+If $queue_label is defined it must be an integer.
+If defined, only tests in the associated queue are considered.
+Otherwise tests from all queues are considered.
+
+Returns the test id and the batch id of the claimed test.
+If there are no waiting tests to claim, C<undef> is returned for both ids.
+
+Only tests in the "waiting" state are considered.
+When a test is claimed it is removed from the queue and it transitions to the
+"running" state.
+
+It is safe for multiple callers running in parallel to allocate tests from the
+same queues.
+
+Dies when an error occurs in the database interface.
+
+=cut
+
 sub get_test_request {
     my ( $self, $queue_label ) = @_;
 
-    my $result_id;
-    my $dbh = $self->dbh;
+    while ( 1 ) {
 
-    my ( $hash_id, $batch_id );
-    if ( defined $queue_label ) {
-        ( $hash_id, $batch_id ) = $dbh->selectrow_array( qq[ SELECT hash_id, batch_id FROM test_results WHERE progress=0 AND queue=? ORDER BY priority DESC, id ASC LIMIT 1 ], undef, $queue_label );
-    }
-    else {
-        ( $hash_id, $batch_id ) = $dbh->selectrow_array( q[ SELECT hash_id, batch_id FROM test_results WHERE progress=0 ORDER BY priority DESC, id ASC LIMIT 1 ] );
-    }
+        # Identify a candidate for allocation ...
+        my ( $hash_id, $batch_id );
```

Comment:
In `claim_test()` and `test_state()` you call it $test_id, here $hash_id. I think the code would be easier to read if it was consistent.

Reply:
I considered renaming it but then I'd rather rename it to $job_id. So I opted to defer the renaming until we make the move to "job".